WordPress/gutenberg
# Cover: Video - Add position controls

![focal-demo](https://user-images.githubusercontent.com/2322354/82595085-05daef80-9b73-11ea-9a2c-d2560084b371.gif)

This update adds positioning controls for video backgrounds within the **Cover block**. Previous to this image, the positioning controls (powered by the `FocalPointPicker` component) were available only to image backgrounds.

## ✨ New Features

##### Click to reposition

The position values can be updated by clicking anywhere in the preview area.

##### Keyboard arrow key support

![Screen Capture on 2020-05-21 at 14-53-25](https://user-images.githubusercontent.com/2322354/82595317-65d19600-9b73-11ea-8ed2-a5210db75ed6.gif)

Up, Down, Left, Right keyboard keys will adjust the position values.

##### Hold Shift to jump by 10

Holding <kbd>Shift</kbd> will jump the values, rounded to the nearest `10`. This works for boht dragging and keyboard arrow inputs.

##### New InputControls

The `top` and `left` values use a new `NumberControl` component. This component includes many features, such as dragging to update, jumping values by holding <kbd>Shift</kbd>, and automatic number rounding.

##### Grid

A 9x9 grid appears anytime a value is updated by **any** interaction (drag, click, etc...)

## 🛠 Fixes

* Enable `<FocalPointPicker />` to reset values based on undo actions in the editor.


## How has this been tested?

This has been tested locally in Storybook and local Gutenberg.

To test...

* Pull down the branch
* Run `npm install`
* Run `npm run dev` to fire up local Gutenberg
* Add a Cover block
* Add a video background (Here's a [sample video to download](https://www.w3schools.com/html/mov_bbb.mp4))
* Adjust the position
* Ensure it works in the editor and on the front-end of the site


## Types of changes

* Add video preview support to `FocalPointPicker`
* Substantial refactors to `FocalPointPicker`
* Adds `Flex` component to `@wordpress/components` as a layout-based primitive
* Update Cover video position rendering in `edit.js` and `save.js`

## Checklist:

Resolves: https://github.com/WordPress/gutenberg/issues/22509
